File: leetcode/Matrix/0054_Spiral_Matrix.cpp
#include <algorithm>
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> result;
        if (matrix.size() == 0 || matrix[0].size() == 0)
            return result;

        vector<vector<bool>> checked(matrix.size(), vector<bool>(matrix[0].size(), false));
        int curR = 0, curC = 0, m = matrix.size(), n = matrix[0].size(), count = 0;
        int direction = 0; // 0: right, 1: down, 2: left, 3: up
        while (count < m * n) {
            result.push_back(matrix[curR][curC]);
            checked[curR][curC] = true;
            count ++;
            switch(direction) {
                case 0:
                    if (curC == n - 1 || checked[curR][curC + 1] == true) {
                        curR ++;
                        direction = 1;
                    }
                    else
                        curC ++;
                    break;
                case 1:
                    if (curR == m - 1 || checked[curR + 1][curC] == true) {
                        curC --;
                        direction = 2;
                    }
                    else
                        curR ++;
                    break;
                case 2:
                    if (curC == 0 || checked[curR][curC - 1] == true) {
                        curR --;
                        direction = 3;
                    }
                    else
                        curC --;
                    break;
                case 3:
                    if (curR == 0 || checked[curR - 1][curC] == true) {
                        curC ++;
                        direction = 0;
                    }
                    else
                        curR --;
                    break;
            }
        }
        return result;
    }
};

File: leetcode/Matrix/0054_Spiral_Matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        N = len(matrix)
        M = len(matrix[0])
        start_i, start_j = 0, 0
        end_i, end_j = N - 1, M - 1
        result = []
        while start_i <= end_i and start_j <= end_j:
            if start_i == end_i:
                # only one row rectangle
                for j in range(start_j, end_j + 1):
                    result.append(matrix[start_i][j])
                start_i += 1
                end_j -= 1
            elif start_j == end_j:
                # only one column rectangle
                for i in range(start_i, end_i + 1):
                    result.append(matrix[i][start_j])
                start_j += 1
                end_j -= 1
            else:
                # print up row
                for j in range(start_j, end_j + 1):
                    result.append(matrix[start_i][j])
                # print right column
                for i in range(start_i + 1, end_i):
                    result.append(matrix[i][end_j])
                # print bottom row
                for j in range(end_j, start_j - 1, -1):
                    result.append(matrix[end_i][j])
                # print left column
                for i in range(end_i - 1, start_i, -1):
                    result.append(matrix[i][start_j])
                start_i += 1
                end_i -= 1
                start_j += 1
                end_j -= 1

        return result
